### ISSUE TITLE
ICE in https://github.com/remram44/rs-sync

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#3717]
````
git clone https://github.com/remram44/rs-sync
git checkout 452a2351a039fdfc68e3386187ae1a57620687a4
RUSTC_BACKTRACE=Full cargo clippy -j 1
````

````
   --> src/delta.rs:209:47
    |
209 |                 if try!(file.read(&mut buffer[idx..idx + 1])) == 0 {
    |                                               ^^^^^^^^^^^^ help: use: `idx..=idx`
    |
    = note: #[warn(clippy::range_plus_one)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#range_plus_one

error: internal compiler error: src/librustc/ty/context.rs:528: node_id_to_type: no type for node `expr <io::Error>::new(io::ErrorKind::InvalidData, "Invalid delta file") (id=2218)`

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:589:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.33.0-nightly (bf669d1e3 2019-01-25) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental -C target-cpu=native --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `rs-sync`.

To learn more, run the command again with --verbose.
````

`clippy 0.0.212 (f175352 2019-01-24)`

**Comment by flip1995** [comment]
Minimal reproducer:
```rust
use std::collections::HashSet;
use std::io;

fn main() {}

pub fn ice_3717(_: &HashSet<usize>) -> io::Result<()> {
    let b = &[0u8; 6];
    if b == b"Clippy" {
        return Err(io::Error::new(io::ErrorKind::InvalidData, ""))
    }
    
    Ok(())
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e5d4cf10781c00cb3acd049dbb20fe8e)

This lint doesn't produce an ICE, when changing the types of the comparison in the if condition. At least I wasn't able to produce the ICE with another `if` condition.

The ICE occurs, because [node_type()](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TypeckTables.html#method.node_type) can't get type informations for `io::Error::new()`.  The weird thing is, that this information depends on the `if` condition.
